// discord-analytics/utils/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractionType = exports.ErrorCodes = exports.ApiEndpoints = void 0;
exports.ApiEndpoints = {
    BASE_URL: 'https://discordanalytics.xyz/api',
    EDIT_SETTINGS_URL: '/bots/:id',
    EDIT_STATS_URL: '/bots/:id/stats',
};
exports.ErrorCodes = {
    INVALID_CLIENT_TYPE: 'Invalid client type, please use a valid client.',
    CLIENT_NOT_READY: 'Client is not ready, please start the client first.',
    INVALID_RESPONSE: 'Invalid response from the API, please try again later.',
    INVALID_API_TOKEN: 'Invalid API token, please get one at ' + exports.ApiEndpoints.BASE_URL.split('/api')[0] + ' and try again.',
    DATA_NOT_SENT: "Data cannot be sent to the API, I will try again in a minute.",
    SUSPENDED_BOT: "Your bot has been suspended, please check your mailbox for more information.",
    INSTANCE_NOT_INITIALIZED: "It seem that you didn't initialize your instance. Please check the docs for more informations.",
    INVALID_EVENTS_COUNT: "invalid events count"
};
var InteractionType;
(function (InteractionType) {
    InteractionType[InteractionType["Ping"] = 1] = "Ping";
    InteractionType[InteractionType["ApplicationCommand"] = 2] = "ApplicationCommand";
    InteractionType[InteractionType["MessageComponent"] = 3] = "MessageComponent";
    InteractionType[InteractionType["ApplicationCommandAutocomplete"] = 4] = "ApplicationCommandAutocomplete";
    InteractionType[InteractionType["ModalSubmit"] = 5] = "ModalSubmit";
})(InteractionType || (exports.InteractionType = InteractionType = {}));
